services/ai_inserts.py

```
@@ -0,0 +1,507 @@
+                    return sanitized
+            except AIClientError as e:
+                logger.warning(f"AI generation failed for holistic intro: {e}")
+        return self._fallback_holistic_intro(inputs)
+
+    def suggest_tags(self, date: str, inputs: Dict[str, Any], force_ai: bool = False) -> List[str]:
+        """Suggest relevant tags from content analysis."""
+        cache_key = self._compute_cache_key("tag_suggestions", inputs)
+        cached_result = self._get_cached_result(date, cache_key)
+        if cached_result and not force_ai:
+            try:
+                return eval(cached_result)
```

Ruff (0.12.2)

Use of possibly insecure function; consider using `ast.literal_eval`

(S307)